sqlobject/src/main/java/org/jdbi/v3/sqlobject/config/internal/UseTemplateEngineImpl.java

```diff
@@ -46,24 +48,60 @@ public void configureForType(ConfigRegistry registry, Annotation annotation, Cla
         }
     }
 
-    private TemplateEngine instantiate(Class<? extends TemplateEngine> value,
+    private static TemplateEngine instantiate(Class<? extends TemplateEngine> value,
                                        Class<?> sqlObjectType,
-                                       Method m) throws Exception {
-        try {
-            Constructor<? extends TemplateEngine> noArg = value.getConstructor();
-            return noArg.newInstance();
-        } catch (NoSuchMethodException e) {
+                                       Method m) {
+
+        return Stream.of(tryConstructor(value), tryConstructor(value, sqlObjectType), tryConstructor(value, sqlObjectType, m), tryConstant(value, "INSTANCE"))
+            .map(Supplier::get)
+            .filter(Objects::nonNull)
+            .findFirst()
+            .orElseThrow(() -> new IllegalStateException("Unable to instantiate, no viable constructor or constant for " + value.getName()));
```

Comment:
From my read, the following would be recognized:

```java
public MyTemplateEngine() {}
public MyTemplateEngine(Class<?> sqlObjectType) {}
public MyTemplateEngine(Class<?> sqlObjectType, Method method) {}
public static final MyTemplateEngine INSTANCE = ...
```

I think the `(Class)` and `(Class, Method)` variants are superfluous. The template engine has access to the SQL Object class and method through `ctx.getExtensionMethod()`.

Aside from that, I feel like if a template engine needs to have options (e.g. the StringSubstitutor engine), it should likewise have its own SqlObject annotation to provide those options.

Reply:
Yeah, that was on my mind too. I didn't want to make functional changes without discussion though, I just refactored the existing code. If you give the go for changing the functionality, I'll give it a try. Maybe in a separate PR? I'm a little confused by some of the implementations though, like for StringTemplate. Most of them really. A lot of their code feels somehow redundant or needlessly complex, and I can't help but think there must be a reason it was done that way.